### ISSUE TITLE
Keep-alive messages, what are they for & compression bug

### ISSUE THREAD

**Issue by pm-araujo** [toverux/expresse#4]
Let me start by pointing out a bug I found when applying your proposed to solution using this module with **expressjs/compression**.
You suggest that `res.flush()` be called after each `res.sse.*` call, but keep-alive messages also call the problematic `res.write` function, since this keep-alive logic happens outside of module user's reach we can't call flush after this happens.
This results in the client never receiving heartbeats until connection is closed or at user-level `res.sse.* ; res.flush();` is called.

Secondly, what exactly are keep-alive messages for in this module?
Initially I thought they were a mechanism to be aware of a lost connection to a user, but there is no logic in place to notify module users of this.
So really what are they for?

**Comment by toverux** [comment]
Hello @pm-araujo,

> You suggest that res.flush() be called after each res.sse.* call, but keep-alive messages also call the problematic res.write function, since this keep-alive logic happens outside of module user's reach we can't call flush after this happens.

Excellent point. That's very right, and I will see what I can do. I never tried expresse with compression so I've never experimented that problem by myself.

Do you have an idea for a proper API?

 - Pragmatic way: check for the existence of `res.flush` and call it after every writes, whether they are keepalives or user writes (through `res.sse.*` functions).

 - "Clean" way: give the user the possibility to have a "post write" hook so they can call `res.flush` (or whatever) themselves. We'd have something like that: `sse({ afterWrite: res => res.flush() })`. Not fully satisfied though, but maybe the best option I see.

 - Other?

> Secondly, what exactly are keep-alive messages for in this module?

Old browsers that do not natively support EventStream may close an inactive connection. For example, on IE, you have to install a polyfill and then ensure the connection is kept. Old browsers kill connections that do not receive datas after some timeout. The keepalive prevents that and we're sure to be safe.

If you want, I can add an option to disable keepalives in the case you're only supporting modern browsers.

**Comment by toverux** [comment]
Hello @pm-araujo,

I'm pleased to announce [v2.3.0](https://github.com/toverux/expresse/releases/tag/v2.3.0), which brings support for expressjs/compression (or any other middleware that has `res.flush`). This is an opt-in feature, ExpreSSE won't just call `res.flush()` if it exists, because this is an old, deprecated express method. It is rewritten by expressjs/compression, but this is not the library's role to check that.

The adopted API is finally expressjs/compression-specific:

```typescript
app.use(compression());

app.get('/events', sse({ flushAfterWrite: true }), (req, res: ISseResponse) => {
    res.sse.comment('Welcome! This is a compressed SSE stream.');
});
```

**Comment by pm-araujo** [comment]
Appreciate the fix to the compression problem.
In regards to the keep-alive messages, currently I'm disabling this functionality at route level by:
``` js
app.get('/events', sse(), (req, res) => {
  res.emit('close');
  // do route logic
}
```
The event is captured by expresse and the scheduling stops: https://github.com/toverux/expresse/blob/0907e875a168532ab593224bce8d93a2a777460d/src/sse_handler_middleware.ts#L56

Not entirely sure this doesn't break express/node functionality, so an option to directly disable this as done with the flush flag would be nice.

**Comment by toverux** [comment]
You're welcome :)

> In regards to the keep-alive messages, currently I'm disabling this functionality at route level by: [...]

Nice hack, I like hacks :) But this is risky indeed, maybe express don't need it because this is handled internally/not used, but other middleware may not like that. In any case, this is non-deterministic. Give me a day or two (I lack time these days) to implement an option to disable keepalives too.

Again, thanks for your feedback on your experience!

**Comment by toverux** [comment]
Hey! Just released v2.4.0 with the requested feature (and another one but it does not relates to this issue in any way):

------

<a name="2.4.0"></a>
# [2.4.0](https://github.com/toverux/expresse/compare/v2.3.0...v2.4.0) (2018-02-12)

### Features

* **sse_handler_middleware:** add `flushHeaders` option, and use proper headers flushing technique ([918bacf](https://github.com/toverux/expresse/commit/918bacf))

```ts
export interface ISseMiddlewareOptions {
    // ...

    /**
     * Whether to flush headers immediately or wait for the first res.write().
     *  - Setting it to false can allow you or 3rd-party middlewares to set more headers on the response.
     *  - Setting it to true is useful for debug and tesing the connection, ie. CORS restrictions fail only when headers
     *    are flushed, which may not happen immediately when using SSE (it happens after the first res.write call).
     *
     * @default true
     */
    flushHeaders: boolean;

    // ...
```

* **sse_handler_middleware:** heartbeats can now be disabled ([dd5a628](https://github.com/toverux/expresse/commit/dd5a628))

```diff
export interface ISseMiddlewareOptions {
    /**
     * Determines the interval, in milliseconds, between keep-alive packets (neutral SSE comments).
+     * Pass false to disable heartbeats (ie. you only support modern browsers/native EventSource implementation and
+     * therefore don't need heartbeats to avoid the browser closing an inactive socket).
     *
     * @default 5000
     */
-    keepAliveInterval: number;
+    keepAliveInterval: false | number;
```
